### PR TITLE
Require

### PR DESCRIPTION
**Pull request by sp1tz**
Tested with corresponding branch in EJS ✅
~There's a subtle bug 🐛 that _must_ be fixed before deployment. It's detailed in review below 📝~

### CODE REVIEW

File: frontend/src/secrets.ts [hunk 1]
```diff
@@ -1,2 +1,3 @@
 export const CLD_FN_BASE_URL : string = 'https://us-central1-arjunguha-research-group.cloudfunctions.net/paws/',
+             MODULE_WL_URL : string = '',
```
Comment (sp1tz) [inline frontend/src/secrets.ts]: This obviously needs to be filled in.

**Comment by arjunguha** [comment]
I do not see a description of the bug.

File: frontend/src/sandbox.ts [hunk 2]
```diff
@@ -16,6 +18,21 @@ export function version() {
     };
 }
 
+let whitelistCode: { [key: string]: string } = {};
+(async function(): Promise<{ [key: string]: string }> {
```
Comment (arjunguha) [inline frontend/src/sandbox.ts]: I'm not sure about having this top-level function call. In particular, we don't want to load the editor *at all* if anything in the whitelist fails to load.

This is going to require hacking in React, similar to logging in. I suggest this approach:

- Give the anonymous loader function in the code below a name, e.g., `loadLibraries(index: Index)`
- The [Index](https://github.com/umass-compsci220/Ocelot/blob/master/frontend/src/mainPage.tsx) class should have a flag in its React state called `allLibsLoaded` with three possible values (`"waiting"`, `"ready"`, or `"failed"`).
- In the Index constructor, we call `loadLibraries(this)`.
- When all libraries are loaded, loadLibraries should call `index.setState({ allLibsLoaded: "ready" })`.
- The [render method](https://github.com/umass-compsci220/Ocelot/blob/master/frontend/src/mainPage.tsx#L56) of `Index` should only display `JumboContent` when `state.allLibsLoaded === "ready"`.

To finish off the UI, Index should display either "loading " or "failed (try to refresh)" when loading fails.

File: frontend/src/mainPage.tsx [hunk 3]
```diff
@@ -43,25 +44,67 @@ type WithStylesClasses =
...
-          case 'safari':
-          case 'ios':
-            const { classes } = this.props;
-            return (<JumboContent.JumboContentDefault classes={classes} />);
-          default:
+export class Index extends React.Component<WithStyles<WithStylesClasses>, {
+  allLibsLoaded: LibsLoad
+}> {
+
+  constructor(props: any) {
```
Comment (arjunguha) [inline frontend/src/mainPage.tsx]: I think this `any` should be `WithStyles<WithStylesClasses>`

File: frontend/src/mainPage.tsx [hunk 4]
```diff
@@ -43,25 +44,67 @@ type WithStylesClasses =
...
+    };
+  }
+
+  componentDidMount() {
+    loadLibraries().then(() => {
+      this.setState({
+        allLibsLoaded: 'ready'
+      });
+    }, (err) => {
+      console.error(`Could not load whitelist: ${JSON.stringify(err)}.`);
```
Comment (arjunguha) [inline frontend/src/mainPage.tsx]: I think "Could not load whitelist" is too specific. It should be "Could not load libraries".

**Comment by arjunguha** [comment]
In the hallway, @sp1tz mentioned that there is a minor bug here. What is it?

**Comment by sp1tz** [comment]
>In the hallway, @sp1tz mentioned that there is a minor bug here. What is it?

NVM. Although, we still do need to set the URL in `secrets.ts`,  otherwise things will blow-up as is. Alternatively, we could leave it empty for now and conditionally fetch in `loadLibraries`.

**Comment by arjunguha** [comment]
So, there are no known bugs here?

**Pull request merged**